RecoLocalCalo/HcalRecProducers/plugins/alpaka/Mahi.dev.cc

```diff
@@ -0,0 +1,1470 @@
+                                      bool const applyTimeSlew,
+                                      float const tzeroTimeSlew,
+                                      float const slopeTimeSlew,
+                                      float const tmaxTimeSlew) const {
+          //2nd index = pulse
+          auto const npulses(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[1u]);
+          //3rd index = sample
+          auto const nsamples(alpaka::getWorkDiv<alpaka::Block, alpaka::Elems>(acc)[2u]);
+
+          auto const nchannels = f01HEDigis.size() + f5HBDigis.size() + f3HBDigis.size();
+          auto const nchannelsf015 = f01HEDigis.size() + f5HBDigis.size();
+
+          auto const recoParamView = recoParamsWithPS.recoParamView();
+          auto const pulseShapeView = recoParamsWithPS.pulseShapeView();
+
+          //Loop over all groups of channels
+          for (auto group : uniform_groups_z(acc, nchannels)) {
+            //Loop over each channel
+            for (auto channel : uniform_group_elements_z(acc, group, nchannels)) {
+              auto const gch = channel.global;
+
+              //Loop over pulses
+              for (auto ipulse : independent_group_elements_y(acc, npulses)) {
+                //Loop over sample
+                for (auto sample : independent_group_elements_x(acc, nsamples)) {
+                  // conditions
+                  auto const id = gch < f01HEDigis.size()
+                                      ? f01HEDigis.ids()[gch]
+                                      : (gch < nchannelsf015 ? f5HBDigis.ids()[gch - f01HEDigis.size()]
+                                                             : f3HBDigis.ids()[gch - nchannelsf015]);
+                  auto const deltaT = gch >= f01HEDigis.size() && gch < nchannelsf015 ? timeSigmaHPD : timeSigmaSiPM;
+
+                  // compute hash for this did
+                  // hash needed to convert condition arrays order (based on Topo) into digi arrays order(based on FED)
+                  auto const did = DetId{id};
+                  auto const hashedId =
+                      did.subdetId() == HcalBarrel
+                          ? did2linearIndexHB(
+                                id, mahi.maxDepthHB(), mahi.firstHBRing(), mahi.lastHBRing(), mahi.nEtaHB())
+                          : did2linearIndexHE(id,
+                                              mahi.maxDepthHE(),
+                                              mahi.maxPhiHE(),
+                                              mahi.firstHERing(),
+                                              mahi.lastHERing(),
+                                              mahi.nEtaHE()) +
+                                mahi.offsetForHashes();
+
+                  // conditions based on the hash
+                  auto const recoPulseShapeId = recoParamView.ids()[hashedId];
+                  auto const* acc25nsVec = pulseShapeView.acc25nsVec()[recoPulseShapeId].data();
```

Comment:
Would it simplify the code to use an approach like
```c++
auto const& pulseShape = pulseShapeView[recoPulseShapeId];
auto const& acc25nsVec = pulseShape.acc25nsVec();
...
```
_etc_.

`acc25nsVec` should be a const reference to an `std::array`, which should work fine in device code.

Reply:
Refactored as per the other discussions. 11aa37463e89e26d17c31a9e9d76e6f799b8a298